api/responses/unordered_countries.php

```diff
@@ -0,0 +1,72 @@
+<?php
+/*
+    Copyright (C) 2004-2010 Kestas J. Kuliukas / Timothy Jones
+
+	This file is part of webDiplomacy.
+
+    webDiplomacy is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    webDiplomacy is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with webDiplomacy.  If not, see <http://www.gnu.org/licenses/>.
+ */
+namespace webdiplomacy_api;
+
+defined('IN_CODE') or die('This script can not be run by itself.');
+
+require_once ('game_country.php');
+
+/**
+ * Games with missing orders JSON response
+ * @package webdiplomacy_api
+ */
+class UnorderedCountries {
+	/**
+	 * List of (gameID, countryID)
+	 * @var array
+	 */
+	public $value = array();
+
+	/**
+	 * Load the countries with missing orders for a given user;
+	 */
+	function load($userID)
+	{
+		global $DB;
+
+		$countryTabl = $DB->sql_tabl("SELECT m.gameID, m.countryID
+                                      FROM wD_Members AS m
+                                      LEFT JOIN wD_Games AS g ON ( g.id = m.gameID )
+                                      WHERE (m.orderStatus IS NULL OR m.orderStatus = '')
+                                            AND m.userID = $userID
+                                            AND g.variantID in (1, 15, 23)
```

Comment:
Can this be configurable please as above

Reply:
I'm not as clear on what data this query is trying to pull, if you explain it in English I can confirm if the right columns are in play here.

Reply:
It lists the game a user is currently playing that have no orders.
This way, the bot can join a game, find that it needs to submit orders for the current phase, and then wait until it needs to submit orders again in the next phase.

SELECT m.gameID, m.countryID
--> Returns the gameID and countryID where the orders as missing

FROM wD_Members AS m
LEFT JOIN wD_Games AS g ON ( g.id = m.gameID )
--> Using the order status from the members table

WHERE (m.orderStatus IS NULL OR m.orderStatus = '')
--> where the orders have not been submitted yet

AND m.userID = $userID
--> JOIN member table on game table

AND g.variantID in (1, 15, 23)
--> Only for variants that are supported by the API

AND g.processStatus = 'Not-processing'
AND g.phase IN ('Diplomacy', 'Retreats', 'Builds');");
--> Only for active games

Reply:
Okay gotcha, so even if you take over for a user, you might not be taking over for all of their games. If a user has a 10 day phase and a 1 day phase game. They can be in CD in the 1 day phase game with another 6 days before they'll even be at risk of CD'ing in the 10 day phase one, even if they don't have orders entered. So you'll want to add in AND m.status = 'Left'

Reply:
There is probably still some confusion here.

There are two components:

1) Bot temporary submitting orders when a player has not submitted their orders in a phase

- SQL query in file "members_in_cd.php"
- Orders can be submitted by bot if no orders have been submitted by a player and there is less than 60 secs for a player to submit their orders before the next process.

2) Bot is playing a game as a regular user (e.g. 1v1 or tournament)

- SQL query in file "unordered_countries.php"
- In this case, we find all the games where the bot user ID is playing that have no orders submitted yet
- There is no need to add "m.status = Left" since we are playing as a regular user.

Reply:
Okay, now I actually gotcha. So in this case for the members_in_cd, my comments about adjusting based on the new excused missed turns making status = 'Left' the correct method are still correct. For the unordered_countries file disregard my comments. Thanks for clarifying the intent between the two.